### DISCUSSION TITLE
Single-run Backfill Cloud Credits

### DISCUSSION THREAD

**Discussion by j-blackwell** [dagster-io/dagster#18963] [Q&A]
If I have a partitioned asset with 10,000 partitions and a backfill policy of `single_run`, how many credits will it cost to backfill this asset in a single run?

1 or 10,000?
The website is a bit unclear as there is only one operation that records 10,000 materialisations.

![image](https://github.com/dagster-io/dagster/assets/33688964/6a101d34-5aa7-480d-a6a4-3f3d88081ad1)

**Comment by gibsondan** [comment]
Each materialization uses up a credit, and each op uses up a credit - so a single-run backfill that materializes N assets will use N + 1 credits (N materializations and 1 op that does all N materializations)

**Reply by j-blackwell** [reply]
> a single-run backfill that materializes N assets will use N + 1 credits

So here I have one asset (10000 partitions), so N = 1. Therefore it is 1 + 1 = 2 credits?

**Reply by j-blackwell** [reply]
The dagster insights have now been published and it is ~ 10000 + 1 credits.

![image](https://github.com/dagster-io/dagster/assets/33688964/af9c64b3-ce59-410b-a59f-47e6467f8235)

**Reply by j-blackwell** [reply]
Combined with this issue https://github.com/dagster-io/dagster/issues/18789, this is *incredibly* dissappointing. @gibsondan , I would really urge the dagster cloud team to revisit this allocation of credits for `single_run` backfill jobs. I cannot see the justification for an 8 second computation to use 1/3 of the monthly allocation of the `Team` plan's credits.

**Reply by frs171717** [reply]
Hi @gibsondan, I just wanted to echo @jrstats's sentiment. Our team has run into the same issue: using many small partitions in the elegant, asset-first way Dagster recommends has led to a mismatch between our marginal cost for additional credits and Dagster’s. We are running a hybrid setup so that mismatch is particularly stark. I hope Dagster considers revisiting the pricing structure as it currently makes it price prohibitive to take advantage of Dagster's partitions feature (which we love and would be happy to continue paying for at a more reasonable cost).

**Reply by wintersrd** [reply]
I also need to share this sentiment; if each partition is billed individually it renders the usage of them prohibitive even on relatively small data. Our simple example:
- We have three tables and need to reload 75 partitions on each per day (these are small data sets compiled in a single API call, but are more usable for our end users if each is a separate file)
- These are three ops per day in total, each <2 minutes run time. Yet as dagster bills, this would be 6840 (6750 partitions, 90 ops) credits

This forces us either to not use partitions or to migrate off cloud to use the platform as designed

**Reply by zyd14** [reply]
Agreed that this is quite the rough edge in the pricing. We were considering adopting partitions more for a number of use cases but will be reconsidering that plan as it would quickly blow out our enterprise plan. Seems like using partitions for data of any medium+ scale data is somewhat prohibitive in cost. Tens or hundreds of thousands of partitions is pretty small in the big data world; we have dozens of assets that would easily have tens or hundreds of thousands of partitions, the cost of modeling this with Dagster partitions far outweighs the value proposition.

**Reply by ion-elgreco** [reply]
I don't expect to ever use the enterprise features with this pricing model in place

**Reply by toddy86** [reply]
+1. I've mentioned this in the Slack threads before. It promotes anti-patterns to "work around" expensive backfill operations. I have a similar concern with the asset checks, which are apparently not being charged for (but by the sounds of it they will be). 

![image](https://github.com/dagster-io/dagster/assets/10559757/14cae603-eab7-41c1-94c2-5de6c6807a81)

Whilst Dagster Cloud is currently an affordable alternative to what we were using previously (MWAA), will all of these additional costs, it's quickly adding up and will very quickly exceed what we were paying. That said, we get a lot more value out of Dagster Cloud (branch deployment etc). So IMO, I don't mind paying the same or a bit more than e.g. MWAA. But if the price keeps escalating, it'll get to a point where it's not affordable for our small operation and we'll probably end up changing to OSS (which I am not a fan of, as I believe in supporting the providers of OSS).

**Reply by ion-elgreco** [reply]
@gibsondan why not just take the log to determine the pricing of a partitioned asset. log (partitions * 10)

So 10 partitions is 2 credits, 100 partitions 3 credits and so on. And you could say the starting price is always 2 credits for partitioned runs when it's above 2 partitions.. This would be way more friendly to do huge backfills, and ongoing partitioned materializations you will still pay extra for using the partitions which is a nice functionality.

**Reply by noahtf13** [reply]
Cannot agree more here. The pricing structure incentivizes you to not use the features as you should. We are very excited about auto-materilization, granular partitions, etc and now we are incentivized to use clunky async jobs and do local backfills because of this weird pricing structure.

I also can't imagine that Dagster's own costs scale with materializations but instead the actual compute time.  

This is an amazing tool but pricing just makes it so frustrating.

**Reply by eliot1785** [reply]
Surprised to see that it works this way, I think you ought to reconsider - there are probably ways to approach this that better reflect the value a customer is getting than backfilled partitions.
